infrastructures/infrastructure-aws-ec2/src/main/java/org/ow2/proactive/resourcemanager/nodesource/infrastructure/AWSEC2Infrastructure.java

```diff
@@ -378,12 +383,46 @@ public void removeNode(Node node) throws RMException {
             logger.warn(e);
         }
 
+        logger.info("Node name :" + node.getNodeInformation().getName() + " InstanceId :" + instanceId);
+
         unregisterNodeAndRemoveInstanceIfNeeded(instanceId,
                                                 node.getNodeInformation().getName(),
                                                 getInfrastructureId(),
                                                 true);
     }
 
+    @Override
+    protected void unregisterNodeAndRemoveInstanceIfNeeded(final String instanceTag, final String nodeName,
+            final String infrastructureId, final boolean terminateInstanceIfEmpty) {
+        setPersistedInfraVariable(() -> {
+            // First read from the runtime variables map
+            //noinspection unchecked
+            nodesPerInstance = (Map<String, Set<String>>) persistedInfraVariables.get(NODES_PER_INSTANCES_KEY);
+            // Make modifications to the nodesPerInstance map
+            if (nodesPerInstance.get(instanceTag) != null) {
+                nodesPerInstance.get(instanceTag).remove(nodeName);
+                logger.info("Removed node: " + nodeName);
+                if (nodesPerInstance.get(instanceTag).isEmpty()) {
+                    logger.info("Instance :" + instanceTag + " is empty ");
+                    if (terminateInstanceIfEmpty) {
+                        logger.info("Call terminate instance for: " + instanceTag);
+                        connectorIaasController.terminateInstance(infrastructureId, instanceTag);
```

Comment:
I think you should use `terminateInstanceByTag` as you are using "instanceTag" returned by `getInstanceIdProperty`.